### PR TITLE
feature/246-router-navigation-guards

### LINKED ISSUE #246: Implement router navigation guards
**Is your feature request related to a problem? Please describe.**
Regardless of being logged in or not a user can enter almost every route without authentication.

**DoD**
- The correct navigation buttons are displayed on the page depending on the user.
- Trying to enter a route you are not supposed to should either take you back or request the user to log in/change account.

**Additional context**
Needs to be discussed with PO regarding what the roles are supposed to be able to do.
Reliant on #237

### PR DESCRIPTION
**Pull request by te4-hugo-svanberg**
## What issue did you implement or fix?
Fixes #246 

## Summary
- Added router navigation guards.
- Added preserveState in UsersModule.

## Type of change
- [x] New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## How can this implementation be improved?
- Add more guards.

### CODE REVIEW

File: frontend/src/router.ts [hunk 1]
```diff
@@ -21,6 +22,13 @@ export default new Router({
       path: '/books/:id',
       name: 'edit book',
       component: views.EditBook,
+      beforeEnter: (to, from, next) => { //eslint-disable-line @typescript-eslint/explicit-function-return-type
+        if (RoleChecker.checkPermission(2)) {
```
Comment (TE4-arvid-nilber) [inline frontend/src/router.ts]: Maybe some comments to explain the number here?

Comment (te4-hugo-svanberg) [reply frontend/src/router.ts]: maybe is the key.

File: frontend/src/helpers/RoleChecker.ts [hunk 2]
```diff
@@ -14,9 +14,9 @@ const roles: Roles = {
 
 export default class RoleChecker {
 
-  private static checkPermission(required: number): boolean {
+  static checkPermission(required: number): boolean {
```
Comment (TE4-jacob-lundberg) [inline frontend/src/helpers/RoleChecker.ts]: Why was this changed? The private was so that you would be forced to use the public methods below to make it easier...

File: frontend/src/helpers/RoleChecker.ts [hunk 3]
```diff
@@ -14,9 +14,9 @@ const roles: Roles = {
...
     const user: User = UsersModule.currentUser;
-    return user && ((user.role & required) == required || (user.role > required));
+    return !!user && ((user.role & required) == required || (user.role > required));
```
Comment (TE4-jacob-lundberg) [inline frontend/src/helpers/RoleChecker.ts]: Do think this is unnecessary due to user being truthy by itself?

File: frontend/src/router.ts (hunk 1 above)
Comment (TE4-jacob-lundberg) [inline frontend/src/router.ts]: Maybe use the already existing methods to check for clearance level?

File: frontend/src/router.ts [hunk 4]
```diff
@@ -32,13 +40,19 @@ export default new Router({
         {path: 'titles', component: views.AdminTitlesView},
         {path: 'books', component: views.AdminBooksView},
         {path: 'cards', component: views.AdminCardsView},
-
-      ]
+      ],
+      beforeEnter: (to, from, next) => { //eslint-disable-line @typescript-eslint/explicit-function-return-type        
+        if (RoleChecker.checkPermission(2)) {
```
Comment (TE4-jacob-lundberg) [inline frontend/src/router.ts]: Same as above

File: frontend/src/helpers/RoleChecker.ts (hunk 2 above)
Comment (te4-hugo-svanberg) [reply frontend/src/helpers/RoleChecker.ts]: Beacuse if the user isn't logged in you got through the 'isStudent' thing.

File: frontend/src/helpers/RoleChecker.ts (hunk 3 above)
Comment (te4-hugo-svanberg) [reply frontend/src/helpers/RoleChecker.ts]: It just returned undefined and crashed otherwise....

File: frontend/src/router.ts (hunk 1 above)
Comment (te4-hugo-svanberg) [reply frontend/src/router.ts]: Mentioned above!

File: frontend/src/router.ts (hunk 4 above)
Comment (te4-hugo-svanberg) [reply frontend/src/router.ts]: Same as above :)

File: frontend/src/helpers/RoleChecker.ts (hunk 2 above)
Comment (TE4-jacob-lundberg) [reply frontend/src/helpers/RoleChecker.ts]: I believe it would be better to modify the several is* and checkPermission to take a user as possible argument. Also getting through isStudent must be because localstorage is not working as it should be. It uses currentUser to check...

**Comment by Linkaan** [comment]
I believe the unit tests fail because the patch for perserveState in isn't working correctly, did you ever verify the implementation @TE4-Mattis-Abrahamsson ?

I've created an issue for this: https://github.com/itggot-TE4/vuex-module-decorators/issues/1

**Comment by TE4-Mattis-Abrahamsson** [comment]
> I believe the unit tests fail because the patch for perserveState in vuex-module-decorators isn't working correctly, did you ever verify the implementation @TE4-Mattis-Abrahamsson ?
> 
> I've created an issue for this: [itggot-TE4/vuex-module-decorators#1](https://github.com/itggot-TE4/vuex-module-decorators/issues/1)

Preserve state is known to cause issues and should not be used in its current state

**Comment by te4-gabriel-stark** [comment]
The login stopped working in this branch for some reason so we will take down this PR.

**Comment by itggot-linus-styren** [comment]
Note for future, awaiting patch in original library or alternative solution that doesn't break existing functionality.
